cmd/snap-confine/snap-confine.c

```diff
@@ -330,6 +295,9 @@ static void enter_non_classic_execution_environment(sc_invocation * inv,
 	struct sc_mount_ns *group = NULL;
 	group = sc_open_mount_ns(inv->snap_instance);
 
+	/* Apply fallback behaviors, if any apply. */
+	sc_apply_invocation_fallback(inv);
```

Comment:
we want to compute is_normal_mode before applying the fallback (that's true also for core16 falling back to core, because we want that to pivot in the end)!

Reply:
Hmm

In my opinion the fallback should be "something is equivalent to something else". The computation of normal vs legacy mode should not be affected by any such replacements.

If we want core16 to differ we should make it sufficient for the system to differentiate core16 from plain core (as in SC_DISTRO_CORE16) so that it can make the sufficient choice by itself.

Otherwise if you actually have core16 installed (thus no fallback to core) will it still use the desired pivot behaviour?

Reply:
I don't understand the reply,  after  fallback from core16 to core the will chose to behave like core so no pivot, exactly what we don't want

Reply:
Two points:

1. https://github.com/snapcore/snapd/commit/44e59fbd243b7f64d23a47a87e40c0c977274bdf#diff-0c384f3cd817f18339705204e8e7b788R304 should not depend on https://github.com/snapcore/snapd/commit/44e59fbd243b7f64d23a47a87e40c0c977274bdf#diff-0c384f3cd817f18339705204e8e7b788R299
2. this patch doesn't implement core16 specific behaviour - it only moves the existing ubuntu-core / core transition logic.

Reply:
I wonder if sc_apply_invocation_fallback() is not the correct name for this function. All it is doing is determining when to change inv->base_snap_name to "ubuntu-core". The logic for what to do with that update is elsewhere. I kind of wonder if we are introducing the potential for a subtle side effect where we have changed inv->base_snap_name to "ubuntu-core" but later (later as in a future update to a later place in the code) we want to use the original.... Perhaps it would be better to introduce another string in the sc_invocation struct, "fallback_base_snap_name" (or similar) that sc_apply_invocation_fallback() updates and then that is what is consulted when determining when to pivot. By doing this, we can call sc_apply_invocation_fallback() really anywhere before the decision to pivot without worry of side effects. This might clarify the correctness of the logic (which I think is what @pedronis (and I for that matter) were struggling with)).

Reply:
> In my opinion the fallback should be "something is equivalent to something else". The computation of normal vs legacy mode should not be affected by any such replacements.

that can be read in two ways,  to be clear my opinion remains that:

*  original base "core" (that means also no base) can be replaced/fulfilled by "ubuntu-core",   and there is no pivot on core 16
*  original base "core16" can be replaced/fulfilled  by "core", but there is no pivoting anywhere

(As we discussed that might mean that if "core16" gets made available there might be a base transition in terms of ns preservation, like if there were a base revision update)

Reply:
I like this suggestion @jdstrand - I will implement something like that.

Reply:
I'm not sure that helps,  almost everything should use the fallback?  unless the idea is to have an  `original` field instead?